cmd/tsidp/tsidp.go

```diff
@@ -1243,3 +1276,80 @@ func isFunnelRequest(r *http.Request) bool {
 	}
 	return false
 }
+
+// isKubeStatePath evaluates whether the provided state path indicates that
+// tailscaled state should be stored in a Kubernetes Secret.
+func isKubeStatePath(statePath string) bool {
+	return strings.HasPrefix(statePath, "kube:")
+}
+
+// validateKubePermissions validates that a tsidp instance has the right
+// permissions to modify its state Secret.
+// It needs to have permissions to get, update and patch the Secret.
+// If the Secret does not already exist, it also needs to have permissions to create it.
+func validateKubePermissions(state string) error {
+	secretName, ok := strings.CutPrefix(state, "kube:")
+	if !ok || secretName == "" {
+		return fmt.Errorf("unable to retrieve valid Kubernetes Secret name from %q", state)
+	}
+
+	kc, err := kubeclient.New("tailscale-tsidp")
+	if err != nil {
+		return fmt.Errorf("error initializing kube client: %w", err)
+	}
+
+	// Our kube client connects to kube API server via the kubernetes
+	// Service in the default namespace, which is not the default client-go
+	// etc behaviour and causes issues to some users. The client defaults
+	// probably cannot be changed for backwards compatibility reasons, but
+	// we can do the right thing here at the same time as adding support for
+	// tsidp to be deployed to kube.
+	url, err := kubeAPIServerAddress()
+	if err != nil {
+		return fmt.Errorf("error initiating kube client: %w", err)
+	}
+	kc.SetURL(url)
+
+	// CheckSecretPermissions returns an error if the permissions to get or
+	// update the Secret are missing. We also require that patch permission
+	// is present and, if the Secret does not exist, create permission.
+	canPatch, canCreate, err := kc.CheckSecretPermissions(context.Background(), secretName)
+	if err != nil {
+		return fmt.Errorf("error checking permissions for Secret %q: %w", secretName, err)
+	}
+	_, err = kc.GetSecret(context.Background(), secretName)
```

Comment:
pass this context in as an arg to validateKubePermissions instead. The caller has one in scope anyway.